### DISCUSSION TITLE
I created a locker that uses Gorm and want to share it.

### DISCUSSION THREAD

**Discussion by manuelarte** [go-co-op/gocron#529] [Show and tell]
Hi,

This is just a locker I created that fulfills our needs, it was not built with the purpose of making it very generic, but I guess it could, potentially inspire other people that could have the same issue. In our case we are using **Postgres and Gorm**. 

In our case, we have 4 pods running, in 2 different deployments and **we are running one job every 24 hours**. For that, every time a pod pick a job, it's storing it in the database, so if another pod tries to run the job, it first check:
- Is there a row in the database with the same key and "time" (trucated to minute), if so, the locking fails.
- If there is not, but I fail with a unique constraint condition, then the locking also fails.
If those conditions are met, a new entry in the database with that information plus the RUNNING status is created, and the unlock method, makes the status FINISHED.

Here is the code:

```
package config

import (
	"context"
	"fmt"
	"time"

	"github.com/go-co-op/gocron"
	"github.com/google/uuid"
	"github.com/sirupsen/logrus"
	"gorm.io/gorm"
)

// This is how the database table will look like, we first need to create it, in our case we use gomigrate for the database migration.
// JobName and JobIdentifier are the unique key.
type CronJobLocker struct {
	ID            uuid.UUID
	CreatedAt     time.Time
	UpdatedAt     time.Time
	JobName       string
	JobIdentifier string
	Worker        string
	Status        string
}

func (c *CronJobLocker) BeforeCreate(_ *gorm.DB) error {
	if c.ID == uuid.Nil {
		c.ID = uuid.New()
	}
	return nil
}


func NewGormJobRunTimeLocker(db *gorm.DB, worker string) gocron.Locker {
	return &gormJobRunTimeLocker{
		Db:     db,
		Worker: worker,
	}
}

var _ gocron.Locker = new(gormJobRunTimeLocker)
type gormJobRunTimeLocker struct {
	Db     *gorm.DB
	Worker string
}

func (p *gormJobRunTimeLocker) Lock(_ context.Context, key string) (gocron.Lock, error) {
	jobIdentifier := time.Now().Truncate(time.Minute).Format("2006-01-02 15:04:05") // we are using the running time of the job to check is uniqueness
	if err := p.isJobRunning(key, jobIdentifier); err != nil {
		return nil, err
	}
	cjb := &CronJobLocker{
		JobName:       key,
		JobIdentifier: jobIdentifier,
		Worker:        p.Worker,
		Status:        "RUNNING",
	}
	tx := p.Db.Create(cjb)
	if tx.Error != nil {
		return nil, tx.Error
	}
	return &gormRunTimeLock{id: cjb.ID, db: p.Db}, nil
}

func (p *gormJobRunTimeLocker) isJobRunning(key string, jobIdentifier string) error {
	cjb := &CronJobLocker{}
	tx := p.Db.Where(&CronJobLocker{JobName: key, JobIdentifier: jobIdentifier}).Find(cjb)
	if tx.Error != nil {
		return tx.Error
	}
	if cjb.ID != uuid.Nil {
		logrus.WithField("worker", cjb.Worker).Info("Job already running by a different worker")
		return fmt.Errorf("job already running by worker %s", cjb.Worker)
	}
	return nil
}

var _ gocron.Lock = new(gormRunTimeLock)

type gormRunTimeLock struct {
	db *gorm.DB
	id uuid.UUID
}

func (p *gormRunTimeLock) Unlock(_ context.Context) error {
	return p.db.Model(&CronJobLocker{ID: p.id}).Updates(&CronJobLocker{Status: "FINISHED"}).Error
}

```

I hope it can help, and please I am open for feedback.

**Comment by JohnRoesler** [comment]
If you would like we could look at making another repo like https://github.com/go-co-op/gocron-redis-lock for gorm and having an implementation there to satisfy the gocron lock interface.

**Comment by manuelarte** [comment]
That would be amazing, how can we proceed?

**Comment by JohnRoesler** [comment]
@manuelarte Here is the repo: https://github.com/go-co-op/gocron-gorm-lock

I'd suggest modeling after the patterns in https://github.com/go-co-op/gocron-redis-lock:
- using test containers go for a docker db implementation for testing https://golang.testcontainers.org/modules/postgres/
- having a docker compose file with a db for manually testing

**Reply by manuelarte** [reply]
Hi @JohnRoesler ,

Unfortunately I did not have so much time to work on this, since I was busy at work. But, I had some time in the weekend to "propose" a first implementation that we may discuss:

https://github.com/go-co-op/gocron-gorm-lock/pull/1

Any feedback is appreciated. I go on holidays so I may not be able to respond/apply comments soon.

Cheers.
